src/cascadia/TerminalCore/ControlKeyStates.hpp

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+namespace Microsoft::Terminal::Core
+{
+    class ControlKeyStates;
+}
+
+constexpr Microsoft::Terminal::Core::ControlKeyStates operator|(Microsoft::Terminal::Core::ControlKeyStates lhs, Microsoft::Terminal::Core::ControlKeyStates rhs) noexcept;
+
+// This class is functionally equivalent to PowerShell's System.Management.Automation.Host.ControlKeyStates enum:
+//   https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.host.controlkeystates
+// It's flagging values are compatible to those used by the NT console subsystem (<um/wincon.h>),
+// as these are being used throughout older parts of this project.
+class Microsoft::Terminal::Core::ControlKeyStates
+{
+    struct StaticValue

```

Comment:
My intention was to have a public `explicit ControlKeyStates(DWORD)` constructor.
At the same time I wanted the enum values below to be `constexpr`.

The only solution for these two requirements I was able to come up with, was to have a "marker type" (here: `StaticValue`) for the enum values, that chooses the correct, public, non-explicit constructor for implicit conversions.
That way you now can do e.g.:
```cpp
constexpr auto v = ControlKeyStates::RightAltPressed | ControlKeyStates::LeftAltPressed;
```

Reply:
What benefit do we gain from marking that constructor as `explicit`? My c++-foo isn't the best, but it seems to me like the only real thing we gain is that you _cant_ call `ControlKeyStates foo = LEFT_ALT_PRESSED;` or `ControlKeyStates foo = 15;`.

I think I'd be happier if `StaticValue` wasn't exposed at all.

Again, I'm not a c++ wizard, but would it be possible to have a private _non_-`explicit` `DWORD` ctor, and a public `explicit` one? Then, use the private one to initialize all the `RightAltPressed` (etc) contants?

Reply:
@zadjii-msft Regarding the potential question of "Why explicit at all?"
I wanted it to prevent accidental integer to `ControlKeyStates` conversions and prevent bugs like this:
```cpp
int generator();
void consumer(ControlKeyStates states);
consumer(generator());
```

The `generator` function might be _completely_ unrelated to any form of key states, which would make this a programming error. You could catch this mistake in a review, but catching it with the help of the compiler seems to me like a free & efficient "linter" on top of it.
More extensive discussions about this can be found anywhere, but a good starting point is [Google's C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Implicit_Conversions)

---

Now regarding...
> but would it be possible to have a private non-explicit DWORD ctor, and a public explicit one

You mean like this?
```cpp
class ControlKeyStates {
public:
    explicit ControlKeyStates(DWORD) {}
private:
    ControlKeyStates(DWORD) {}
};
```

That's not possible. You can't have two functions (or constructors) with the same signature, but different visibilities or modifiers.

But what I _intended_ to write was actually similar to what you suggested and something you might've liked:
```cpp
class ControlKeyStates {
public:
    explicit ControlKeyStates(DWORD) {}

    static constexpr ControlKeyStates RightAltPressed{ RIGHT_ALT_PRESSED };
    // ...
};
```

But alas that's also not possible because when you instantiate a variable the type must be fully defined, which is only the case _after_ the class-statement is finished.

GCC and Clang allow you to write the following valid C++17 code, but compile it as a proper constexpr (which isn't defined in the language spec):
```cpp
class ControlKeyStates {
public:
    explicit ControlKeyStates(DWORD) {}

    static const ControlKeyStates RightAltPressed{ RIGHT_ALT_PRESSED };
    // ...
};

static constexpr ControlKeyStates ControlKeyStates::RightAltPressed{ RIGHT_ALT_PRESSED };
// ...
```

But MSVC fails to recognize the constexpr and thus fails to link such code.
So you can't have a private constructor with the same signature but different visibility and you can't define constexpr members outside of the class definition in MSVC.
The only other solution is to have a public "marker type" that has access to your private, "insecure" constructor.
Since this will only be a simple marker type (here: a struct with one field) it can then also be fully defined the moment you want to instantiate the constexpr members within the ControlKeyStates class definition. 🙂

I hope all of that made any sense to you. 😅

Reply:
That actually does make a lot more sense. Thanks for taking the time to write it up!